src/nnet3/nnet-general-component.h

```diff
@@ -440,6 +440,142 @@ class StatisticsPoolingComponentPrecomputedIndexes:
+  }
+
+  virtual void ZeroStats();
+
+  virtual Component* Copy() const {
+    return new BackpropTruncationComponent(dim_,
+                                     clipping_threshold_,
+                                     zeroing_threshold_,
+                                     zeroing_interval_,
+                                     recurrence_interval_,
+                                     num_clipped_,
+                                     num_zeroed_,
+                                     count_,
+                                     count_zeroing_boundaries_);}
+
+  virtual void Propagate(const ComponentPrecomputedIndexes *indexes,
+                         const CuMatrixBase<BaseFloat> &in,
+                         CuMatrixBase<BaseFloat> *out) const;
+  virtual void Backprop(const std::string &debug_info,
+                        const ComponentPrecomputedIndexes *indexes,
+                        const CuMatrixBase<BaseFloat> &, // in_value,
+                        const CuMatrixBase<BaseFloat> &, // out_value,
+                        const CuMatrixBase<BaseFloat> &out_deriv,
+                        Component *to_update,
+                        CuMatrixBase<BaseFloat> *in_deriv) const;
+
+  virtual ComponentPrecomputedIndexes* PrecomputeIndexes(
+      const MiscComputationInfo &misc_info,
+      const std::vector<Index> &input_indexes,
+      const std::vector<Index> &output_indexes,
+      bool need_backprop) const;
+
+  virtual void Scale(BaseFloat scale);
+  virtual void Add(BaseFloat alpha, const Component &other);
+  virtual void Read(std::istream &is, bool binary); // This Read function
+  // requires that the Component has the correct type.
+  /// Write component to stream
+  virtual void Write(std::ostream &os, bool binary) const;
+  virtual std::string Info() const;
+  virtual ~BackpropTruncationComponent() {
+  }
+ private:
+  int32 dim_;  // input/output dimension
+  
+  BaseFloat clipping_threshold_;  // threshold to be used for clipping
+                                  // corresponds to max-row-norm
+
+  BaseFloat zeroing_threshold_;   // threshold to be used for zeroing
+                                  // corresponds to max-row-norm
+
+  int32 zeroing_interval_;        // the interval in number of frames that we
+                                  // would apply the zeroing of the gradient
+
+  int32 recurrence_interval_;     // together with zeroing_interval_ to decide
+                                  // the frame at which for zeroing the gradient
+
+  std::string debug_info_;   // component-node name, used in the destructor to
+                             // print out stats of self-repair
+
+  BackpropTruncationComponent &operator =
+      (const BackpropTruncationComponent &other); // Disallow.
+
+ protected:
+  // variables to store stats
+  // An element corresponds to rows of derivative matrix
+  int32 num_clipped_;  // number of elements which were clipped
```

Comment:
These types should be double, otherwise operations like Add() and Scale() may give weird effects.